onboard/catkin_ws/src/task_planning/scripts/combination_tasks.py

```diff
@@ -0,0 +1,125 @@
+
+from task import Task
+
+
+class ListTask(Task):
+    """Run a list of tasks sequentially"""
+
+    def __init__(self, tasks, num_loops=1):
+        super(ListTask, self).__init__()
+
+        self.num_loops = num_loops
+        self.tasks = tasks
+        self.curr_index = 0
+
+    def _on_task_run(self):
+        if self.curr_index == len(self.tasks):
+            if self.num_loops > 0:
+                self.num_loops -= 1
+                if self.num_loops == 0:
+                    self.finish()
+
+            self.curr_index = 0
+
+        elif self.tasks[self.curr_index].finished:
+            if self.num_loops != 1:
+                self.tasks[self.curr_index].restart()
+            self.curr_index += 1
+
+        else:
+            self.tasks[self.curr_index].run()
+
+
+class IndSimulTask(Task):
+    """Run a list of tasks simulataneously, exits when all tasks finished"""
+
+    def __init__(self, tasks):
+        super(IndSimulTask, self).__init__()
+
+        self.tasks = tasks
+
+    def _on_task_run(self):
+        self.all_finished = True
+        for task in self.tasks:
+            task.run()
+            if not task.finished:
+                self.all_finished = False
+
+        if self.all_finished:
+            self.finish()
+    
+    def restart(self):
+        self.unfinish()
+        for task in self.tasks:
+            task.restart()
+        self._on_task_start()
+
+
+class DepSimulTask(Task):
+    """Run a list of tasks simulataneously, exits when any task finished"""
+
+    def __init__(self, tasks):
+        super(DepSimulTask, self).__init__()
+
+        self.tasks = tasks
+
+    def _on_task_run(self):
+        self.any_finished = False
+        for task in self.tasks:
+            task.run()
+            if task.finished:
+                self.any_finished = True
+
+        if self.any_finished:
+            for task in self.tasks:
+                task.finish()
+            self.finish()
+
+    def restart(self):
+        self.unfinish()
+        for task in self.tasks:
+            task.restart()
+        self._on_task_start()
+
+
+class LeaderFollowerTask(Task):
+
+    def __init__(self, leader, follower):
+        super(LeaderFollowerTask, self).__init__()
+        self.leader = leader
+        self.follower = follower
+
+    def _on_task_run(self):
+        if self.leader.finished:
+            self.follower.finish()
+            self.finish()
+
+        self.leader.run()
+        self.follower.run()
+
+    def restart(self):
+        self.unfinish()
+        self.leader.restart()
+        self.follower.restart()
+        self._on_task_start()
+
+
+class IfElseTask(Task):
```

Comment:
This task doesn't make sense to me, condition seems to be a variable, so if you know what the value of the variable is when creating the task, why wouldn't you assign the task there, instead of pushing it onto this task.

I think this would make more sense if condition is a function, something that can be evaluated, so you can evaluate it during start, instead of during initialization.

Reply:
We will be re-evaluating this while switching to the new system. It'll probably make sense to remove it.